doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,653 @@
+also available.
+
+It is important to note that `BIO_dgram_get_local_addr_cap` is entirely distinct
+from the application capability negotiation mechanism discussed above. Whereas
+the capability negotiation mechanism discussed above allows *applications* to
+signal what they are capable of handling in their usage of a given BIO,
+`BIO_dgram_local_addr_cap` allows a *BIO implementation* to indicate to the
+users of that BIO whether it is able to support local addressing (where
+enabled).
+
+#### `BIO_s_dgram_pair`
+
+**See also:**
+[BIO_s_dgram_pair(3)](https://www.openssl.org/docs/manmaster/man3/BIO_s_dgram_pair.html)
+
+A new BIO implementation, `BIO_s_dgram_pair`, is provided. This is similar to
+the existing BIO pair but provides datagram semantics. It provides full support
+for the new APIs `BIO_sendmmsg`, `BIO_recvmmsg`, the capability negotiation
+mechanism described above, local address support and the MTU signalling
+mechanism described above.
+
+It can be instantiated using the new API `BIO_new_dgram_pair`.
+
+#### `BIO_POLL_DESCRIPTOR`
+
+The concept of *poll descriptors* are introduced. A poll descriptor is a tagged
+union structure which represents an abstraction over some unspecified kind of OS
+descriptor which can be used for synchronization and waiting.
+
+The most commonly used kind of poll descriptor is one which describes a network
+socket (i.e., on POSIX-like platforms, a file descriptor), however other kinds
+of poll descriptor may be defined.
+
+A BIO may be queried for whether it has a poll descriptor for read or write
+operations respectively:
+
+- Where `BIO_get_rpoll_descriptor` (`BIO_CTRL_GET_RPOLL_DESCRIPTOR`) is called,
+  the BIO should output a poll descriptor which describes a resource which can
+  be used to determine when the BIO will next become readable via a call to
+  `BIO_read` or, if supported by the BIO, `BIO_recvmmsg`.
+- Where
+  `BIO_get_wpoll_descriptor` (`BIO_CTRL_GET_WPOLL_DESCRIPTOR`) is called, the
+  BIO should output a poll descriptor which describes a resource which can be
+  used to determine when the BIO will next become writeable via a call to
+  `BIO_write` or, if supported by the BIO, `BIO_sendmmsg`.
+
+A BIO may not necessarily be able to provide a poll descriptor. For example,
+memory-based BIOs such as `BIO_s_dgram_pair` do not correspond to any OS
+synchronisation resource, and thus the `BIO_get_rpoll_descriptor` and
+`BIO_get_wpoll_descriptor` calls are not supported for such BIOs.
+
+A BIO which supports these functions is known as pollable, and a BIO which does
+not is known as non-pollable. `BIO_s_dgram` supports these functions.
+
+The implementation of these functions for a `BIO_f_ssl` forwards to
+`SSL_get_rpoll_descriptor` and `SSL_get_wpoll_descriptor` respectively. The
+
+#### `BIO_s_dgram_mem`
+
+This is a basic memory buffer BIO with datagram semantics. Unlike
+`BIO_s_dgram_pair`, it is unidirectional and does not support peer addressing or
+local addressing.
```

Comment:
It should be extended to support this (as well as support for the various MTU ctrls)